Use DEVICE_DT_GET to obtain chosen entropy device

gmarull opened the pull request:
A reference to the CHOSEN entropy device can be obtained at compile time, so
avoid using device_get_binding(). As a result of the changes, `DT_CHOSEN_ZEPHYR_ENTROPY_LABEL` is no longer needed and so has been dropped.

Resolves https://github.com/zephyrproject-rtos/zephyr/issues/43230

gmarull committed:
kernel: use DEVICE_DT_GET_OR_NULL for entropy device

A reference to the entropy device can be obtained at compile time, so
avoid using device_get_binding().

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
modules: hal_nordic: obtain entropy device using DEVICE_DT_GET

The entropy device can be obtained at compile time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
bluetooth: controller: ll_sw: obtain entropy device using DEVICE_DT_GET

The entropy device can be obtained at compile time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
net: lib: openthread: obtain entropy device using DEVICE_DT_GET

The device can be obtained at compile time. Note that now the readiness
is checked on every call, but it's a fast operation anyway.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
random: rand32_ctr_drbg: obtain entropy device using DEVICE_DT_GET

The entropy device can be obtained at compile time. Adjusted the
assertion message slightly, as it no longer uses the entropy macro
label. Also adjusted the entropy_driver variable to entropy_dev, since
it does not hold a "driver" but a "device".

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
random: rand32_entropy_device: get entropy device using DEVICE_DT_GET

THe entropy device can be ontained at compile time. Simplify some
logs/code as a result. Also adjusted "entropy_driver" name to
"entropy_dev", since it holds a "device", not a "driver".

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
random: rand32_xoshiro128: obtain entropy device using DEVICE_DT_GET

The entropy device reference can be obtained at compile time. Adjust
error code to -ENODEV.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: clock_control: nrf_onoff_and_bt: use DEVICE_DT_GET

Use DEVICE_DT_GET to obtain a reference to the chosen entropy device.
The device is now global, and readiness is checked at the test setup
fixture.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: drivers: entropy: api: use DEVICE_DT_GET

Obtain a reference to the chosen entropy device at compile time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

mbolivar-nordic commented on kernel/init.c:
```diff
@@ -363,10 +363,10 @@ void z_early_boot_rand_get(uint8_t *buf, size_t length)
 	int n = sizeof(uint32_t);
 #ifdef CONFIG_ENTROPY_HAS_DRIVER
-	const struct device *entropy = device_get_binding(DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
+	const struct device *entropy = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
I think you mean `DEVICE_DT_GET_OR(DT_CHOSEN(zephyr_entropy), NULL);`, right?

gmarull replied:
fixed

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll.c:
```diff
@@ -52,7 +52,7 @@ static struct {
 
 /* Entropy device */
-static const struct device *dev_entropy;
+static const struct device *dev_entropy = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
Should lack of an entropy device really be a build error here too?

gmarull replied:
It looks like the module is not operational without this device, so I think so. Looks like `init_reset	` error code is propagated up to the BT HCI open() method. Maybe I'm wrong, will defer judgement to maintainers.

mbolivar-nordic replied:
@dleach02 can you please chime in here, or help us find the right person to make a call about this?

mbolivar-nordic commented on subsys/net/lib/openthread/platform/entropy.c:
```diff
@@ -19,22 +19,19 @@ LOG_MODULE_REGISTER(net_otPlat_entropy, CONFIG_OPENTHREAD_L2_LOG_LEVEL);
 #error OpenThread requires an entropy source for a TRNG
 #endif
 
+static const struct device *dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
Should this be a build error, for sure?

gmarull replied:
Module is not operational without this device, whether it's a failure that's handled later: I don't know. We need maintainers voice here.

rlubos replied:
I think it's ok to require at build time to have an entropy device present in the system, AFAIK there's no fallback mechanism in OT for that. @edmont What do you think?

edmont replied:
I agree.

mbolivar-nordic commented on subsys/random/rand32_ctr_drbg.c:
```diff
@@ -51,16 +51,11 @@ static int ctr_drbg_initialize(void)
-			DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
-		return -EINVAL;
+	if (!device_is_ready(entropy_driver)) {
+		__ASSERT(0, "Device driver for %s not ready. "
```
It would be nice to see this changed to "Entropy device %s not ready". We have a legacy in zephyr of mixing up drivers with devices, and calling `struct device` instances "drivers", which I think is a misnomer that would be nice to improve wherever we can while we're "modernizing".

gmarull replied:
adjusted, pls check

mbolivar-nordic commented on subsys/random/rand32_entropy_device.c:
```diff
@@ -9,28 +9,19 @@
-		entropy_driver = dev;
-	}
+	__ASSERT(device_is_ready(entropy_driver),
+		 "Device driver for %s not ready. "
```
Same here.

mbolivar-nordic commented on subsys/random/rand32_entropy_device.c:
```diff
@@ -9,28 +9,19 @@
 #include <string.h>
 
-static const struct device *entropy_driver;
+static const struct device *entropy_driver = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
entropy_device?

mbolivar-nordic commented on subsys/random/rand32_entropy_device.c:
```diff
@@ -48,23 +39,14 @@ uint32_t z_impl_sys_rand32_get(void)
-		entropy_driver = dev;
-	}
+	__ASSERT(device_is_ready(entropy_driver),
+		 "Device driver for %s not ready. "
```
ditto here

mbolivar-nordic commented on subsys/random/rand32_xoshiro128.c:
```diff
@@ -41,9 +41,9 @@ static inline uint32_t rotl(const uint32_t x, int k)
-	dev = device_get_binding(DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
-	if (!dev) {
-		return -EINVAL;
+	dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
Are you sure this should be a build error?

gmarull replied:
Without this device, it appears that this rand source is not usable, so I think it should be. Will let maintainers judge here.

mbolivar-nordic replied:
> Will let maintainers judge here.

Can you `@`-mention the relevant maintainers, since this PR touches different subsystems?

ceolin replied:
without an entropy source this module can't operate. It is fine for me to have a build error.

mbolivar-nordic commented on include/devicetree/zephyr.h:
```diff
@@ -47,10 +40,6 @@
 
 #endif /* DT_DOXYGEN */
 
-#if DT_NODE_HAS_PROP(DT_CHOSEN(zephyr_entropy), label)
-#define DT_CHOSEN_ZEPHYR_ENTROPY_LABEL DT_LABEL(DT_CHOSEN(zephyr_entropy))
-#endif
```
This needs to go through the stable API change process.

gmarull replied:
https://github.com/zephyrproject-rtos/zephyr/issues/43230

gmarull committed:
devicetree: deprecate DT_CHOSEN_ZEPHYR_ENTROPY_LABEL

Deprecate the macro in favor of DT_LABEL(DT_CHOSEN(zephyr_entropy)), and
promote usage of DEVICE_DT_GET(), where labels are not required.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin reviewed (approved):
Changes look good to me.

gmarull commented:
@mbolivar-nordic ping

dleach02 commented on subsys/random/rand32_ctr_drbg.c:
```diff
@@ -28,7 +28,7 @@
 static K_SEM_DEFINE(state_sem, 1, 1);
 
-static const struct device *entropy_driver;
+static const struct device *entropy_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_entropy));
```
unfortunately I missed this "chime" request... there is a bug that was introduced with this change because the initialization function is only called based on whether this entropy_dev object has been assigned (see line 105 below).
